upup/pkg/fi/assettasks/copyfile.go

```diff
@@ -77,75 +98,65 @@ func (_ *CopyFile) Render(c *fi.Context, a, e, changes *CopyFile) error {
-	sourceSha := fi.StringValue(e.SourceSha)
-	sourceSHALocation := fi.StringValue(e.SourceShaLocation)
+	sourceSha := fi.StringValue(e.SHA)
 
-	glog.Infof("copying bits from %q to %q", source, target)
+	glog.V(2).Infof("copying bits from %q to %q", source, target)
 
-	if err := transferFile(c, source, target, sourceSha, sourceSHALocation); err != nil {
+	if err := transferFile(c, source, target, sourceSha); err != nil {
 		return fmt.Errorf("unable to transfer %q to %q: %v", source, target, err)
 	}
 
 	return nil
 }
 
-func transferFile(c *fi.Context, source string, target string, sourceSHA string, sourceSHALocation string) error {
+// transferFile downloads a file from the source location, validates the file matches the SHA,
+// and uploads the file to the target location.
+func transferFile(c *fi.Context, source string, target string, sha string) error {
+
+	// TODO drop file to disk, as vfs reads file into memory.  We load kubelet into memory for instance.
+	// TODO in s3 can we do a copy file ... would need to test
+
 	data, err := vfs.Context.ReadFile(source)
 	if err != nil {
-		return fmt.Errorf("Error unable to read path %q: %v", source, err)
+		return fmt.Errorf("error reading path %q: %v", source, err)
 	}
 
 	uploadVFS, err := buildVFSPath(target)
 	if err != nil {
 		return err
 	}
 
+	shaVFS, err := buildVFSPath(target + ".sha1")
+	if err != nil {
+		return err
+	}
+
 	// Test the file matches the sourceSha
 	{
-		var sha string
-		if sourceSHA != "" {
-			sha = sourceSHA
-		} else if sourceSHALocation != "" {
-			shaBytes, err := vfs.Context.ReadFile(sourceSHALocation)
-			if err != nil {
-				return fmt.Errorf("Error unable to read path %q: %v", source, err)
-			}
-			sha = string(shaBytes)
-		}
-
-		if sha != "" {
-
-			trimmedSHA := strings.TrimSpace(sha)
-
-			in := bytes.NewReader(data)
-			dataHash, err := hashing.HashAlgorithmSHA1.Hash(in)
-			if err != nil {
-				return fmt.Errorf("unable to hash sha from data: %v", err)
-			}
 
-			shaHash, err := hashing.FromString(trimmedSHA)
-			if err != nil {
-				return fmt.Errorf("unable to hash sha: %q, %v", sha, err)
-			}
+		in := bytes.NewReader(data)
+		dataHash, err := hashing.HashAlgorithmSHA1.Hash(in)
+		if err != nil {
+			return fmt.Errorf("unable to hash sha from data: %v", err)
+		}
 
-			if !shaHash.Equal(dataHash) {
-				return fmt.Errorf("SHAs are not matching for %q", dataHash.String())
-			}
+		shaHash, err := hashing.FromString(strings.TrimSpace(sha))
+		if err != nil {
+			return fmt.Errorf("unable to hash sha: %q, %v", sha, err)
+		}
 
-			shaVFS, err := buildVFSPath(target + ".sha1")
-			if err != nil {
-				return err
-			}
+		if !shaHash.Equal(dataHash) {
+			return fmt.Errorf("SHAs are not matching for %q", dataHash.String())
+		}
 
-			b := bytes.NewBufferString(sha)
-			if err := writeFile(c, shaVFS, b.Bytes()); err != nil {
-				return fmt.Errorf("Error uploading file %q: %v", shaVFS, err)
-			}
+		b := []byte(shaHash.Hex())
```

Comment:
You want to write the file before you upload the SHA, unless you're going to always hash the dest file.